[Azure Pipelines] Upgrade to macOS 11 (Big Sur)

foolip opened the pull request.

foolip committed:
[Azure Pipelines] Upgrade STP runs to macOS 11 (Big Sur)

This keeps Safari (stable) on Catalina for now, due to lack of
support for acceptInsecureCerts which we're relying on for Big Sur.

Additionally, this moves all usage of Python on Big Sur to 3.7, as 3.6
is not available Azure Pipelines' macOS 11 VMs.

Co-authored-by: Sam Sneddon <gsnedders@apple.com>

foolip commented:
This did not go as smoothly as I had hoped. There have apparently been changes to the security model for installing certificates, leading to this failure:

> SecTrustSettingsSetTrustSettings: errSecInternalComponent

@gsnedders do you know if it's possible to work around this in any way?

gsnedders commented:
I'm pretty sure the problem is that `/Library/Keychains/System.keychain` is on a non-writeable volume as of Big Sur; adding it to the user keychain is probably the right solution here.

foolip commented:
Thanks, I'll look into that!

foolip commented:
@gsnedders I tried some things but no luck. The latest is this:
```sh
KEYCHAIN=$(security default-keychain -d user)
echo $KEYCHAIN
KEYCHAIN=$(eval echo $KEYCHAIN)
echo $KEYCHAIN
sudo security add-trusted-cert -r trustRoot -k "$KEYCHAIN" tools/certs/cacert.pem
```

The `echo`s confirmed the path is /Users/runner/Library/Keychains/login.keychain-db. In other words, trying to add the cert to the user keychain but using `sudo` to avoid the password prompt. But that doesn't work, the error is still "SecTrustSettingsSetTrustSettings: errSecInternalComponent".

This feels very similar to past issues of `sudo safaridriver --enable` not working. Do you have some ideas for other things to try? Perhaps it's possible to tell Safari (STP) which keychain (or cert directly) to use in some other way, as command line arguments or via WebDriver?

stephenmcgruer commented:
On my local macbook, doing:

```
sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain tools/certs/cacert.pem
OR
sudo security add-trusted-cert -r trustRoot -k /Users/smcgruer/Library/Keychains/login.keychain-db tools/certs/cacert.pem
```

Both prompt me for my sudo password on the cmdline, **and** then open a UX dialog explaining I am making changes to the 'System Certificate Trust Settings' and that I need to enter my password to allow it...

https://github.com/travis-ci/travis-ci/issues/5789 mentions this (long before Big Sur?). It ends up pointing to https://www.jviotti.com/2016/03/16/how-to-code-sign-os-x-electron-apps-in-travis-ci.html which creates its own keychain instead of using the system one. That doesn't necessarily feel like the right solution though.

Another forum post I found, with a lot of sysadmins complaining about this change in Big Sur :D -  https://developer.apple.com/forums/thread/671582

Some folks have also noted that you can 'echo' the password into the cmd, but this requires putting it in the clear. (And for azure pipelines we likely don't have that password anyway...)

foolip commented:
Thanks for digging @stephenmcgruer! Indeed we don't have the password in this case, or at least when I was dealing with the `safaridriver --enable` issue I couldn't get it from the Azure Pipelines team.

gsnedders commented:
The only way to do add certificates without explicit user consent is via disabling SIP.

stephenmcgruer commented:
Relevant documentation from Apple: https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection (which feels fairly infeasible...)

thejohnjansen commented:
I do not think it is possible. We haven't spent a lot of time investigating work-arounds, but the design implies this is not possible.

gsnedders committed:
Use acceptInsecureCerts with a new enough release of Safari

This removes the blocker to running STP on Azure Pipelines documented
in #31147, whereby trusting certificates in Big Sur or later requires
either user interaction or SIP being disabled, which is infeasible in
public automation.

gsnedders committed:
Make installing certs only happen on Azure Pipelines on macOS < 11

See
https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-release-notes#Security
for details about the behaviour change in Big Sur; due to the
requirement for user interaction this is infeasible on Azure
Pipelines.

gsnedders committed:
Update Safari Technology Preview to 135

gsnedders commented:
@foolip this appears to currently be failing due to chromedriver on both Linux and macOS, `unknown error (500): unknown error: cannot find Chrome binary`, which is pretty unfortunate

https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=73539 has a full run of STP underway, and hopefully that comes out looking okay

foolip commented:
That run completed: https://wpt.fyi/results/?run_id=5714547575881728

foolip commented:
https://wpt.fyi/results/?run_id=5638166280142848 is the closest run to compare to, so:
https://wpt.fyi/results/?diff&filter=ADC&run_id=5638166280142848&run_id=5714547575881728

Overall, definitely more fixes than regressions.

@gsnedders I guess only passing the CI now blocks this? Is "cannot find Chrome binary" also happening on master? I'll revive my probing PR to see...

jgraham commented on tools/ci/azure/install_certs.yml:
```diff
@@ -1,7 +1,8 @@
 - script: |
     # https://github.com/web-platform-tests/results-collection/blob/master/src/scripts/trust-root-ca.sh
-    sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain tools/certs/cacert.pem
+    # only run this on macOS < 11
```
Couldn't we just remove this whole step?

gsnedders replied:
we're still running Safari Stable on macOS 10.15, and we don't have `acceptInsecureCerts` support in stable

foolip commented:
Azure Pipelines and Taskcluster were happy in https://github.com/web-platform-tests/wpt/pull/28759, so failures here are probably introduced by the changes.

foolip reviewed (commented):
I can't approve since I started the PR, but current changes (sans the `./wpt test-jobs --all` bits) look good!

gsnedders committed:
Update infrastructure expectations inline with Azure Pipelines results
